### DISCUSSION TITLE
Help with Micro SD Card and ILI9341 TFT Display

### DISCUSSION THREAD

**Discussion by Mark25A** [micropython/micropython#14502] [ESP32] [answered]
Hi, I've been struggling to combine an External Micro SD Card with the ILI9341 TFT display.

I am working on a project for my college project, I needed 5 components along with the ESP32 dev kit on MicroPyhton. All my components work well, and in fact, my project is finished. But only recently I got my hands on the micro sd card. For a few days I tried to combine it with the display, but it didn't work.

For testing I got some help from ChatGPT, which helped, but we are stuck in a loop, and I can't even get help from my teachers, because they have no clue either...

My final exam is in less than a month away.

I tried running the code that finally allowed me to read a text file. But the tft screen is loading a black image, and when I press the push button to communicate with the sd card, it reads the text, and along with it it seems to fill a very small portion of the screen with the color white.
I tried turning the TFT's Chip Select high, and the SD's low, and only accessible when pressing the button. I even tried disabling the microsd card for 6 seconds, and still the same issue.

If I remove the initialization of the sd card, the image loads without any issues, but of course I can't access the sd card that way.

Also, for some reason I need to reformat my sd card sometimes after running one of the codes.

Below is one of the codes in Thonny I had saved, I'll appreciate some help.

import machine
from time import sleep
from ili9341 import Display, color565
from machine import Pin, SPI, SoftSPI
import os
from sdcard import SDCard

# Initialize TFT display
spi_display = SPI(2, baudrate=40000000, sck=Pin(18), mosi=Pin(23))
display_cs = Pin(19, Pin.OUT)
display = Display(spi_display, dc=Pin(5), cs=display_cs, rst=Pin(17))

# Initialize SD card (pins)
spisd = SoftSPI(-1, miso=Pin(19), mosi=Pin(23), sck=Pin(18))
sd_cs = Pin(22, Pin.OUT)

# Button setup
button = Pin(33, Pin.IN, Pin.PULL_UP)

#Function to display the intro image
def intro():
    image_path = 'Intro.raw'  # Adjust the path as needed
    try:
        with open(image_path, 'rb') as f:
            display.draw_image(image_path, 0, 0, 240, 320)
    except OSError:
        print("Error: Image file not found")

# def intro():
#     display.draw_image('Intro.raw', 0,0,240,320)

# Function to read the text file from SD card
def read_sd_text_file():
    try:
        sd = SDCard(spisd, sd_cs)
        vfs = os.VfsFat(sd)
        os.mount(vfs, '/sd')
        print('Mounted SD card. Root directory:', os.listdir('/sd'))
        
        os.chdir('/sd/templates')
        with open('test.txt', 'r') as f:
            content = f.read()
            print("Content of the file:")
            print(content)
    except OSError as e:
        print("Error:", e)
    finally:
        os.umount('/sd')  # Unmount SD card
        print('Unmounted SD card.')

# Button interrupt handler
def button_pressed(pin):
    # Deselect TFT display
    display_cs.value(1)
    sleep(0.1)  # Small delay to ensure proper deselection
    
    # Read from SD card
    read_sd_text_file()
    
    # Re-select TFT display
    display_cs.value(0)
    sleep(0.1)  # Small delay to ensure proper selection

# Set up button interrupt
button.irq(trigger=Pin.IRQ_FALLING, handler=button_pressed)

if __name__ == "__main__":
    # Select TFT display
    display_cs.value(0)
    # Display the intro image
    intro()
    print("Intro image displayed. Waiting for button press to read SD card...")

**Comment by Mark25A** [comment]
I noticed that it posted the code badly, so I'll post it as a text file.
[SDCard_And_TFT_TEST.txt](https://github.com/micropython/micropython/files/15354798/SDCard_And_TFT_TEST.txt)

**Comment by Mark25A** [comment]
Forgot to add, the libraries I use are:

TFT Display: https://github.com/rdagger/micropython-ili9341
MicroSD: https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/storage/sdcard/sdcard.py
Video about SD Card: https://www.youtube.com/watch?v=qL2g5YIVick

**Reply by GitHubsSilverBullet** [reply]
> For testing I got some help from ChatGPT, which helped, but we are stuck in a loop, and I can't even get help from my teachers, because they have no clue either...
> My final exam is in less than a month away.

If your exam is a month away and you're depending on – dare I say it – ChatGPT aka. A.S.
then you have far more problems than you can imagine.

Proper formatting: triple backticks followed by py or python.

**Reply by Mark25A** [reply]
Well I figured it might help me in that case but it seems like chatgpt isn't all that good as they say. I thought maybe trying to handle interrupt it, but before I could do that the sd card stopped working. I think the sd card itself was already problematic because it is no longer recognizable on any device, so that might as well have been the issue all along.

**Comment by ricksorensen** [comment]
You have two different SPI drivers defined using same SCK and MOSI- the SoftSPI and the hardware SPI.  Even though they use different CS pins it seems like this could be a conflict.  To test either use two completely different SPI definitions (different SCK, MOSI, MISO, ... baudrate) or use as single SPI and just set the different CS for use.

**Reply by peterhinch** [reply]
@Mark25A It is also bad practice to have `sleep` commands in an interrupt handler. Your teachers will explain the reason.

**Reply by Mark25A** [reply, marked as answer]
Yup, I know the difference now and I'm changing to timers. I didn't know that it delays everything,  in fact, the teachers weren't those that told me about it, it was my Father.

Also, I figured my issue out, although still not perfect.  But I used an interrupt that will initialize the microsd only when i need to.

So my problem is solved. Thanks for the replies.

**Comment by Mark25A** [comment]
> @Mark25A It is also bad practice to have `sleep` commands in an interrupt handler. Your teachers will explain the reason.
